docs/design/2021-04-18-common-table-expression.md

```diff
@@ -0,0 +1,287 @@
+<!--
+This is a template for TiDB's change proposal process, documented [here](./README.md).
+-->
+
+# Proposal: Support Common Table Expression
+
+- Author(s):     [@wjhuang2016](https://github.com/wjhuang2016) [@guo-shaoge](https://github.com/guo-shaoge)
+- Last updated:  2021-04-18
+- Discussion at: https://github.com/pingcap/tidb/issues/17472 <!-- https://github.com/pingcap/tidb/issues/XXX -->
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+* [Future Work](#future-work)
+
+## Introduction
+
+This proposal describes the design of Common Table Expression.
+
+## Motivation or Background
+
+The Common Table Expression(CTE) is a temporary result set which exists within a statement. You can reference it later in the statement.
+It's similar to derived tables in some way. But CTE has extra advantages over derived tables.
+
+1. CTE can be referenced multiple times.
+2. CTE is easier to read.
+3. You can use CTE to do hierarchical queries.
+
+There are two kinds of CTE:
+1. non-recursive CTE.
+
+    ```
+    WITH
+        cte1 AS (SELECT c1 FROM t1)
+        cte2 AS (SELECT c2 FROM t2)
+    SELECT cte1.c1, cte2.c2 FROM cte1, cte2 WHERE cte1.c1 = cte2.c2 AND cte1.c1 = 100;
+    ```
+2. recursive CTE, which can be used to do hierarchical queries. Recursive CTE normally consists of seed part and recursive part.
+   Seed part will generate the origin data, the remaining computation will be done by recursive part.
+
+    ```
+    WITH RECURSIVE cte1 AS (
+        SELECT part, sub_part FROM t WHERE part = 'human'
+        UNION ALL
+        SELECT t.part, t.sub_part FROM t, cte1 WHERE cte1.sub_part = t.part
+    )
+    SELECT * FROM cte1;
+    ```
+
+## Detailed Design
+### Overview
+There are two ways to implement CTE:
+1. Merge: just like view, CTE is expanded on where it's used.
```

Comment:
Don't mention the 'Merge' algorithm here. Which makes the design document complicate.

Reply:
Maybe we should let readers know why we dont support Merge method.
Guess it's better to mention Merge method here?